R/arc-read.R

```diff
@@ -0,0 +1,146 @@
+#'   vector. If `col_names` is one less than the length of the default or if
+#'   `col_names` if `FALSE`, the existing sf column name is retained. If
+#'   `col_names` is the string "alias", names are set to match the available
+#'   alias names for the layer.
+#' @param n_max Maximum number of records to return. Defaults to 10000 or an
+#'   option set with `options("arcgislayers.n_max" = <max records>)`
+#' @inheritParams arc_select
+#' @inheritParams arc_raster
+#' @param fields Fields to return. Ignored if `col_select` is supplied.
+#' @param name_repair See [vctrs::vec_as_names()] for details. Set `name_repair
+#'   = NULL` to avoid using [vctrs::vec_as_names()].
+#' @param ... Additional arguments passed to [arc_select()] if URL is a
+#'   "FeatureLayer" or "Table" or [arc_raster()] if URL is an "ImageLayer".
+#' @examples
+#' if (interactive()) {
+#'   url <- "https://sampleserver6.arcgisonline.com/arcgis/rest/services/Census/MapServer/3"
+#'
+#'   arc_read(url)
+#'
+#'   arc_read(url, name_repair = tolower)
+#'
+#'   url <- "https://sampleserver6.arcgisonline.com/arcgis/rest/services/EmergencyFacilities/FeatureServer/0"
+#'
+#'   arc_read(url, col_names = "alias")
+#'
+#'   img_url <- "https://landsat2.arcgis.com/arcgis/rest/services/Landsat/MS/ImageServer"
+#'
+#'   arc_read(
+#'     img_url,
+#'     width = 1000, height = 1000,
+#'     xmin = -71, ymin = 43,
+#'     xmax = -67, ymax = 47.5,
+#'     bbox_crs = 4326
+#'   )
+#' }
+#' @export
+arc_read <- function(url,
+                     col_names = TRUE,
+                     col_select = NULL,
+                     n_max = getOption("arcgislayers.n_max", default = 10000),
+                     name_repair = "unique",
+                     crs = NULL,
+                     ...,
+                     fields = NULL,
+                     token = Sys.getenv("ARCGIS_TOKEN")) {
+  service <- arc_open(url = url, token = token)
+
+  crs <- crs %||% sf::st_crs(service)
+
+  if (!obj_is_layer(service)) {
+    layer <- arc_raster(
+      x = service,
+      ...,
+      crs = crs,
+      token = token
+    )
+
+    return(layer)
+  }
+
+  if (attr(service, "n") > n_max) {
+    cli::cli_alert_warning(
+      "{.arg n_max} is set to {n_max}, less than the {attr(service, 'n')}
+      features available from this service."
+    )
+  }
+
+  layer <- arc_select(
+    x = service,
+    fields = col_select %||% fields,
+    crs = crs,
+    n_max = n_max,
+    token = token,
+    ...
+  )
+
+  set_layer_names(
+    layer,
+    col_names = col_names,
+    name_repair = name_repair,
+    alias = service[["fields"]][["alias"]]
+  )
+}
+
+#' @noRd
+set_layer_names <- function(x,
+                            col_names = NULL,
+                            name_repair = NULL,
+                            alias = NULL,
+                            call = rlang::caller_env()) {
+  layer_nm <- names(x)
+  nm <- layer_nm
+  sf_column_nm <- attributes(x)[["sf_column"]]
```

Comment:
I think perhaps we want `attr(x, "sf_col")` instead of fetching all of the attributes and subsetting we can get just one.